From 07307749228489a630c126d953ff701de15c7e77 Mon Sep 17 00:00:00 2001
From: Dan Walsh <dwalsh@redhat.com>
Date: Mon, 14 Jul 2014 08:50:01 -0400
Subject: [PATCH] Allow logrotate to manage virt_cache

---
 logrotate.te | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/logrotate.te b/logrotate.te
index 17ea89c..7908e77 100644
--- a/logrotate.te
+++ b/logrotate.te
@@ -292,6 +292,10 @@ optional_policy(`
 	varnishd_manage_log(logrotate_t)
 ')
 
+optional_policy(`
+	virt_manage_cache(logrotate_t)
+')
+
 #######################################
 #
 # logrotate_mail local policy
-- 
2.1.0

